libwild/src/input_data.rs

```diff
@@ -362,7 +362,8 @@ impl<'data> InputData<'data> {
 
     pub(crate) fn has_file(&self, name: &'data [u8]) -> bool {
         self.path_to_load_index
-            .contains_key(Path::new(OsStr::from_bytes(name)))
+            .keys()
+            .any(|path| path.ends_with(Path::new(OsStr::from_bytes(name))))
```

Comment:
It's not the best solution but more correct. Albeit also more costly, hopefully unnoticeably.

I'm not sure if it'll work in every case.

Reply:
Does it matter that before, if a file had a DT_NEEDED entry that was `/foo/libbar.so` and we had loaded `/foo/libbar.so` that we'll now return false here?

I wonder if instead of looking at filenames we should be matching DT_NEEDED with DT_SONAME. i.e. Build a HashSet of the DT_SONAME for all shared objects that we loaded, then look up DT_NEEDED entries in that HashSet.

Reply:
Sorry, I don't understand what you mean. Both previous, and current code will return true for `/foo/libbar.so`: https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=eda40aa773ad36d58404da41a6a6dcfb
Otherwise, `shlib-undefined.c-disallow-complete` test would fail, as it uses absolute path (as noted in the description).

Without digging into it, checking DT_SONAME sounds like a better option than filename. With a fallback to filename if absent.

Reply:
> Sorry, I don't understand what you mean. Both previous, and current code will return true for `/foo/libbar.so`

You're right, I misinterpreted what was happening

Reply:
No worries. I'll merge it as an improvement over previous behaviour, but we certainly want to try out cases where soname != filename.

Reply:
Created the issue https://github.com/davidlattimore/wild/issues/1325